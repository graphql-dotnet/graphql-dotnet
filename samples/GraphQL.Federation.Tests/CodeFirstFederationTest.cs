using System.Text.Json;
using GraphQL.DataLoader;
using GraphQL.Federation.Tests.Fixtures;
using GraphQL.SystemTextJson;

namespace GraphQL.Federation.Tests;

// dotnet test GraphQL.Federation.Tests --filter DisplayName~GraphQL.Federation.Tests.CodeFirstFederationTest.ServiceTest
// dotnet test GraphQL.Federation.Tests --filter DisplayName~GraphQL.Federation.Tests.CodeFirstFederationTest.EntitiesTest
[Collection(nameof(CodeFirstCollectionDefinition))]
public class CodeFirstFederationTest : BaseCodeFirstGraphQLTest
{
    public CodeFirstFederationTest(CodeFirstFixture codeFirstFixture)
        : base(codeFirstFixture)
    { }

    [Fact]
    public async Task ServiceTest()
    {
        // WaitForDebugger();

        var json = await Schema.ExecuteAsync(new GraphQLSerializer(), options =>
        {
            options.Listeners.Add(new DataLoaderDocumentListener(new DataLoaderContextAccessor()));
            options.ThrowOnUnhandledException = true;
            options.Query = """
                query {
                    _service {
                        sdl
                    }
                }
                """;
        });

        var sdl = JsonDocument.Parse(json).RootElement
            .GetProperty("data")
            .GetProperty("_service")
            .GetProperty("sdl")
            .GetString()!;

        sdl.ShouldMatchApproved(c => c.NoDiff());
    }

    [Fact]
    public async Task EntitiesTest()
    {
        // WaitForDebugger();

        var json = await Schema.ExecuteAsync(new GraphQLSerializer(), options =>
        {
            options.Listeners.Add(new DataLoaderDocumentListener(new DataLoaderContextAccessor()));
            options.ThrowOnUnhandledException = true;
            options.Query = """
                query {
                    _entities(representations: [{ __typename: "FederatedTestDto", id: 1 }, { __typename: "FederatedTestDto", id: 3 }, { __typename: "ExternalResolvableTestDto", id: 123, external: "asdfgh" }]) {
                        ... on FederatedTestDto {
                            name
                            externalTest {
                                id
                            }
                            externalResolvableTest {
                                id
                                external
                                extended
                            }
                        }
                        ... on ExternalResolvableTestDto {
                            id
                            extended
                        }
                    }
                }
                """;
        });

        json.ShouldBe(
            """{"data":{"_entities":[{"name":"111","externalTest":{"id":4},"externalResolvableTest":{"id":7,"external":"external-7","extended":"external-7"}},{"name":"333","externalTest":{"id":6},"externalResolvableTest":{"id":9,"external":"external-9","extended":"external-9"}},{"id":123,"extended":"asdfgh"}]}}""",
            StringCompareShould.IgnoreLineEndings);
    }
}
