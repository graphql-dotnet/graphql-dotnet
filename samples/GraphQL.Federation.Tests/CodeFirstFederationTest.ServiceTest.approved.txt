schema @link(url: "https://specs.apollo.dev/link/v1.0", import: ["@link"]) @link(url: "https://specs.apollo.dev/federation/v2.3", import: ["@key", "@external", "@requires", "@provides", "@shareable", "@inaccessible", "@override", "@tag"]) {
  query: TestQuery
}

scalar _Any

type TestQuery {
  directivesTest: DirectivesTestDto!
  _service: _Service!
  _entities(representations: [_Any!]!): [_Entity]!
}

type DirectivesTestDto @key(fields: "id") @shareable @inaccessible {
  id: Int!
  shareable: String! @shareable
  inaccessible: String! @inaccessible
  override: String! @override(from: "OtherSubgraph")
  external: String! @external
  provides: String! @provides(fields: "foo bar")
  requires: String! @requires(fields: "foo bar")
}

type _Service {
  sdl: String
}

union _Entity = DirectivesTestDto | FederatedTestDto | ExternalResolvableTestDto

type FederatedTestDto @key(fields: "id") {
  id: Int!
  name: String! @deprecated(reason: "Test deprecation reason 01.")
  externalTestId: Int!
  externalResolvableTestId: Int!
  externalTest: ExternalTestDto! @deprecated(reason: "Test deprecation reason 02.")
  externalResolvableTest: ExternalResolvableTestDto! @provides(fields: "external")
}

type ExternalTestDto @key(fields: "id", resolvable: false) {
  id: Int!
}

type ExternalResolvableTestDto @key(fields: "id") {
  id: Int!
  external: String! @external
  extended: String! @requires(fields: "External")
}
